File: Leetcode/L_Remove_Nth_Node_From_End_of_List.py
# Remove Nth Node From End of List


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        fast = slow = head
        for i in range(n):
            fast = fast.next
        if fast is None:
            return head.next
        while fast.next:
            fast = fast.next
            slow = slow.next
        slow.next = slow.next.next
        return head


X = Solution()
print(X.removeNthFromEnd([1, 2, 3, 4, 5],  2))

File: Leetcode Js/L_Remove_Nth_Node_From_End_of_List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
  if (!head.next && n > 0) {
    return head.next;
  }
  const length = lengthOfLinkedList(head);

  const nFromStart = length - n + 1;

  let count = 1;
  let prev = null;
  let cur = head;

  if (nFromStart === 1) {
    return head.next;
  }
  while (count !== nFromStart && cur.next) {
    prev = cur;
    cur = cur.next;
    count++;
  }

  prev.next = cur.next;
  return head;
};

const lengthOfLinkedList = (head) => {
  let count = 1;
  let node = head;
  while (node && node.next) {
    node = node.next;
    count++;
  }
  return count;
};
